arch/xtensa/src/esp32s2/esp32s2_oneshot.c

```diff
@@ -0,0 +1,462 @@
+
+      oneshot->running = true;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp32s2_oneshot_cancel
+ *
+ * Description:
+ *   Cancel the oneshot timer and return the time remaining on the timer.
+ *
+ * Input Parameters:
+ *   oneshot Caller allocated instance of the oneshot state structure.  This
+ *           structure must have been previously initialized via a call to
+ *           esp32s2_oneshot_initialize();
+ *   ts      The location in which to return the time remaining on the
+ *           oneshot timer.  A time of zero is returned if the timer is
+ *           not running.  ts may be zero in which case the time remaining
+ *           is not returned.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success.  A call to up_timer_cancel() when
+ *   the timer is not active should also return success; a negated errno
+ *   value is returned on any failure.
+ *
+ ****************************************************************************/
+
+int esp32s2_oneshot_cancel(struct esp32s2_oneshot_s *oneshot,
+                           struct timespec *ts)
+{
+  int ret = OK;
+  uint64_t current_us;
+  uint64_t remaining_us;
+  uint64_t timeout_us;
+  uint64_t counter_value;
+  uint64_t alarm_value;
+
+  DEBUGASSERT(oneshot);
+
+  if (oneshot->running == false)
+    {
+      tmrinfo("Trying to cancel a non started oneshot timer.\n");
+      ts->tv_sec  = 0;
+      ts->tv_nsec = 0;
+    }
+  else
+    {
+      /* Stop timer */
+
+      ESP32S2_TIM_STOP(oneshot->tim);
+
+      /* Disable int */
+
+      ESP32S2_TIM_DISABLEINT(oneshot->tim);
+
+      /* Detach handler */
+
+      ret = ESP32S2_TIM_SETISR(oneshot->tim, NULL, NULL);
+
+      if (ts != NULL)
+        {
+          /* Get the current counter value */
+
+          ESP32S2_TIM_GETCTR(oneshot->tim, &counter_value);
+
+          /* Get the current configured timeout */
+
+          ESP32S2_TIM_GETALRVL(oneshot->tim, &alarm_value);
+
+          current_us = counter_value * oneshot->resolution;
+          timeout_us = alarm_value   * oneshot->resolution;
+
+          /* Remaining time (us) = timeout (us) - current (us) */
+
+          remaining_us = timeout_us - current_us;
+          ts->tv_sec   = remaining_us / USEC_PER_SEC;
+          remaining_us = remaining_us - ts->tv_sec * USEC_PER_SEC;
+          ts->tv_nsec  = remaining_us * NSEC_PER_USEC;
+        }
+
+        oneshot->running  = false;
+        oneshot->handler  = NULL;
+        oneshot->arg      = NULL;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: esp32s2_oneshot_current
+ *
+ * Description:
+ *   Get the current time.
+ *
+ * Input Parameters:
+ *   oneshot Caller allocated instance of the oneshot state structure.  This
+ *           structure must have been previously initialized via a call to
+ *           esp32s2_oneshot_initialize();
+ *   usec    The maximum delay in us.
```

Comment:
```suggestion
 *
 * Output Parameters:
 *   usec    The current time in us.
```
`usec`is an output parameter here. Also, I believe the parameter description might be inaccurate.